### DISCUSSION TITLE
Ignoring DeprectionWarning in pytest.warns

### DISCUSSION THREAD

**Discussion by lafrech** [pytest-dev/pytest#8666] [Q&A] [answered]
The docs read:

> By default, DeprecationWarning and PendingDeprecationWarning will not be caught when using pytest.warns() or recwarn because the default Python warnings filters hide them. If you wish to record them in your own code, use warnings.simplefilter('always').

It doesn't seem to be the case. My tests use `pytest.warns` to record warnings and `DeprecationWarning`s are recorded.

```py
        with pytest.warns(None) as record:
            function_that_calls_a_lib_that_generates_a_deprecation_warning()
        assert not record
```

My understanding of this is limited. Could it be due to changes in Python 3.7 regarding `DeprecationWarning`?

Am I missing something?

The workaround I found for now is to filter the records afterwards, which is verbose and not elegant (managing `PendingDeprecationWarning` as well would be even longer).

```py
        with pytest.warns(None) as record:
            function_that_calls_a_lib_that_generates_a_deprecation_warning()
        assert not [r for r in record if r.category is not DeprecationWarning]
```

------------------------------------------------------

Reproducible test case.

Using Python 3.7.3

```py
import warnings

import pytest


def test_deprecation_warning():


    with pytest.warns(None) as record:
        warnings.warn("Test", DeprecationWarning)

    # This fails
    # assert not record

    print(record)
    # WarningsChecker(record=True)

    print([r.category for r in record])
    # [<class 'DeprecationWarning'>]

    # This succeeds
    assert not [r for r in record if r.category is not DeprecationWarning]
```

At some point, I thought it had something to do with the use of `None` in the call to `warns` but it doesn't:

```py
import warnings

import pytest


def test_deprecation_warning():

    with pytest.warns(RuntimeWarning) as record:
        warnings.warn("Test", DeprecationWarning)
        warnings.warn("my warning", RuntimeWarning)

    print(record)
    # WarningsChecker(record=True)

    print([r.category for r in record])
    # [<class 'DeprecationWarning'>, <class 'RuntimeWarning'>]
```

**Comment by nicoddemus** [comment, marked as answer]
Hmmm strange. Do you have any warning filters active? Can you also provide a reproducible example?

**Reply by lafrech** [reply]
Sorry for the delay.

I added a test case. Fresh virtualenv with latest pytest, no filter. Python 3.7.3.

**Reply by nicoddemus** [reply]
Thanks and sorry about the delay.

Actually the docs are outdated, see https://github.com/pytest-dev/pytest/pull/8754. 👍 

Thanks a lot for bringing this to attention.

**Reply by lafrech** [reply]
Thanks for the update.

And for all the work on pytest.
